azure/services/natgateways/client.go

```diff
@@ -55,42 +62,127 @@ func netNatGatewaysClient(subscriptionID string, baseURI string, authorizer auto
+	if !ok {
+		if params == nil {
+			// nothing to do here.
+			return existingNatGateway, nil, nil
+		}
+		return nil, nil, errors.Errorf("%T is not a network.NatGateway", params)
+	}
+
+	future, err := ac.natgateways.CreateOrUpdate(ctx, spec.ResourceGroupName(), spec.ResourceName(), natGateway)
 	if err != nil {
-		return err
+		return nil, nil, err
 	}
+
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureCallTimeout)
+	defer cancel()
+
 	err = future.WaitForCompletionRef(ctx, ac.natgateways.Client)
 	if err != nil {
-		return err
+		// if an error occurs, return the future.
+		// this means the long-running operation didn't finish in the specified timeout.
+		return nil, &future, err
 	}
-	_, err = future.Result(ac.natgateways)
-	return err
+
+	result, err := future.Result(ac.natgateways)
+
+	// if the operation completed, return a nil future
+	return result, nil, err
 }
 
-// Delete deletes the specified nat gateway.
-func (ac *azureClient) Delete(ctx context.Context, resourceGroupName, natGatewayName string) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "natgateways.AzureClient.Delete")
-	defer done()
+// DeleteAsync deletes a Nat Gateway asynchronously. DeleteAsync sends a DELETE
+// request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
+// progress of the operation.
+func (ac *azureClient) DeleteAsync(ctx context.Context, spec azure.ResourceSpecGetter) (azureautorest.FutureAPI, error) {
+	ctx, span := tele.Tracer().Start(ctx, "natgateways.azureClient.Delete")
+	defer span.End()
 
-	future, err := ac.natgateways.Delete(ctx, resourceGroupName, natGatewayName)
+	future, err := ac.natgateways.Delete(ctx, spec.ResourceGroupName(), spec.ResourceName())
 	if err != nil {
-		return err
+		return nil, err
 	}
+
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureCallTimeout)
+	defer cancel()
+
 	err = future.WaitForCompletionRef(ctx, ac.natgateways.Client)
 	if err != nil {
-		return err
+		// if an error occurs, return the future.
+		// this means the long-running operation didn't finish in the specified timeout.
+		return &future, err
 	}
 	_, err = future.Result(ac.natgateways)
-	return err
+	// if the operation completed, return a nil future.
+	return nil, err
+}
+
+// IsDone returns true if the long-running operation has completed.
+func (ac *azureClient) IsDone(ctx context.Context, future azureautorest.FutureAPI) (bool, error) {
+	ctx, span := tele.Tracer().Start(ctx, "natgateways.azureClient.IsDone")
+	defer span.End()
+
+	done, err := future.DoneWithContext(ctx, ac.natgateways)
+	if err != nil {
+		return false, errors.Wrap(err, "failed checking if the operation was complete")
+	}
+
+	return done, nil
+}
+
+// Result fetches the result of a long-running operation future.
+func (ac *azureClient) Result(ctx context.Context, futureData azureautorest.FutureAPI, futureType string) (interface{}, error) {
```

Comment:
Add `tele.Tracer().Start()` here?

Reply:
@CecileRobertMichon Should we be doing this for Result()? I don't think we've been adding it on the other async services.

Reply:
It should, it has a context. We should also add it to the others, can be a separate commit here or later in a separate PR.

Reply:
actually it should be tele.StartSpanWithLogger